File: music-commons-api/src/main/java/net/transgressoft/commons/music/audio/AudioItem.java
package net.transgressoft.commons.music.audio;

import net.transgressoft.commons.query.QueryEntity;

import java.nio.file.Path;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Set;

/**
 * @author Octavio Calleya
 */
public interface AudioItem extends QueryEntity {

    Path path();

    AudioItem path(Path path);

    String fileName();

    String extension();

    String title();

    AudioItem title(String name);

    Artist artist();

    AudioItem artist(Artist artist);

    Set<String> artistsInvolved();

    Album album();

    AudioItem album(Album album);

    Genre genre();

    AudioItem genre(Genre genre);

    String comments();

    AudioItem comments(String comments);

    short trackNumber();

    AudioItem trackNumber(short trackNumber);

    short discNumber();

    AudioItem discNumber(short discNumber);

    float bpm();

    AudioItem bpm(float bpm);

    Duration duration();

    String encoder();

    AudioItem encoder(String encoder);

    String encoding();

    AudioItem encoding(String encoding);

    long length();

    int bitRate();

    short playCount();

    AudioItem playCount(short playCount);

    LocalDateTime dateOfInclusion();

    LocalDateTime lastDateModified();
}

File: music-commons-api/src/main/kotlin/net/transgressoft/commons/music/audio/AudioItem.kt
package net.transgressoft.commons.music.audio

import net.transgressoft.commons.IdentifiableEntity
import java.nio.file.Path
import java.time.Duration
import java.time.LocalDateTime

interface AudioItem : IdentifiableEntity<Int> {
    val path: Path
    val fileName: String
    val extension: String
    var title: String
    val duration: Duration
    val bitRate: Int
    var artist: Artist
    val artistsInvolved: Set<String>
    var album: Album
    var genre: Genre
    var comments: String?
    var trackNumber: Short?
    var discNumber: Short?
    var bpm: Float?
    val encoder: String?
    val encoding: String?
    val length: Long
    var coverImageBytes: ByteArray?
    val dateOfCreation: LocalDateTime
    val lastDateModified: LocalDateTime

    fun writeMetadata()
}
